source/slang/ir.cpp

```diff
@@ -3737,8 +3794,46 @@ namespace Slang
         context->shared = &sharedContextStorage;
         context->builder = &sharedContextStorage.builderStorage;
         context->target = target;
-
-
+        Substitutions * curTailSubst = nullptr;
+        for (auto param : programLayout->genericEntryPointParams)
+        {
+            auto paramSubst = new GlobalGenericParamSubstitution();
+            if (!context->subst)
+                context->subst = paramSubst;
+            if (curTailSubst)
+                curTailSubst->outer = paramSubst;
+            curTailSubst = paramSubst;
+            paramSubst->paramDecl = param->decl;
+            SLANG_ASSERT((UInt)param->index < entryPointRequest->genericParameterTypes.Count());
+            paramSubst->actualType = entryPointRequest->genericParameterTypes[param->index];
+            // find witness tables
+            for (auto witness : entryPointRequest->genericParameterWitnesses)
+            {
+                if (auto subtypeWitness = witness.As<SubtypeWitness>())
+                {
+                    if (subtypeWitness->sub->EqualsVal(paramSubst->actualType))
```

Comment:
Actually... yeah, this seems kind of broken, because you are having to decide for the witnesses, which witness belongs to which type argument.

What I would kind of expect here is that a `GlobalGenericParameterSubstition` would have something like:

```
List<DeclRef<Decl>> params;
List<RefPtr<Val>> args;
```

And these would basically be copied directly from data filled in by the front-end. The basic approach would be:

- For each global generic type parameter, add a `DeclRef` to it to the `params` array, and then add the actual type to the `args` array at the same index

- For each constraint on a global type parameter, add a `DeclRef` to the constraint to the `params` array, and then add the `SubtypeWitnessVal` to the `args` array at the same index

Then substitution is just a matter of doing a (linear) search for a known `DeclRef` in the `params` array, and if the decl is found, swapping in the corresponding value from the `args` array (at that point this is *almost* exactly like the non-global `GenericSubstition` case).

That representation completely drops the need to organize things based on what constraints belong to what parameters (which looks a bit hacky here). You'd still need to look for `SubtypeWitnessVal` entries in the `args` array and swap them out for an `IRProxyVal` to the corresponding witness table, which seems to be the interesting work going on here.

Reply:
I could have used two lists, but the one thing I was not sure about is what if we have a type that implements multiple interfaces, in that case wouldn't we have many SubtypeWitnessVals for a single parameter? That means we are not going to assume a 1-1 relationship between a parameter and a witness anyways. And that was why I've chosen to implement things this way.